recipe/magento2.php

```diff
@@ -219,3 +221,137 @@
 ]);
 
 after('deploy:failed', 'magento:maintenance:disable');
+
+// artifact deployment section
+// settings section
+set('artifact_file', 'artifact.tar.gz');
+set('artifact_dir', 'artifacts');
+set('artifact_excludes_file', 'artifacts/excludes');
+
+set('artifact_path', function () {
+    if (!test('[ -d {{artifact_dir}} ]')) {
+        run('mkdir {{artifact_dir}}');
+    }
+    return get('artifact_dir') . '/' . get('artifact_file');
+});
+
+set('bin/tar', function () {
+    if (commandExist('gtar')) {
+        return which('gtar');
+    } else {
+        return which('tar');
+    }
+});
+
+set('cacheToolPath', function() {
+    return get('cacheTool', '{{current_path}}/bin/cachetool');
+});
+
+// tasks section
+desc('Packages all relevant files in an artifact.');
+task('artifact:package', function() {
+    if (!test('[ -f {{artifact_excludes_file}} ]')) {
+        throw new GracefulShutdownException(
+            "No artifact excludes file provided, provide one at artivacts/excludes or change location"
+        );
+    }
+    run('{{bin/tar}} --exclude-from={{artifact_excludes_file}} -czf {{artifact_path}} .');
+});
+
+desc('Uploads artifact in release folder for extraction.');
+task('artifact:upload', function () {
+    upload(get('artifact_path'), '{{release_path}}');
+});
+
+desc('Extracts artifact in release path.');
+task('artifact:extract', function () {
+    run('{{bin/tar}} -xzpf {{release_path}}/{{artifact_file}} -C {{release_path}}');
+    run('rm -rf {{release_path}}/{{artifact_file}}');
+});
+
+desc('Provides env.php for build.');
+task('build:prepare-env', function() {
+    $deployEnv = get('deploy_env','app/etc/deploy.php');
+    if (!test('[ -f ./'.$deployEnv.' ]')) {
+        throw new GracefulShutdownException(
+            "No deploy env provided, provide one at app/etc/deploy.php or change location"
+        );
+    }
+    run ('cp '.$deployEnv.' app/etc/env.php');
+});
+
+desc('Clears generated files prior to building.');
+task('build:remove-generated', function() {
+    run('rm -rf generated/*');
+});
+
+desc('Clears the opcache, cache tool required.');
+task('cache:clear:opcache', function() {
+    if ($fpmSocket = get('fpm_socket', '')) {
+        run('{{bin/php}} -f {{cacheToolPath}} opcache:reset --fcgi '.$fpmSocket);
+    }
+});
+
+desc('Builds an artifact.');
+task('artifact:build', function () {
+    if(currentHost()->get('local')) {
+        set('deploy_path', '.');
+        set('release_path', '.');
+        set('current_path', '.');
+        invoke('build:prepare-env');
+        invoke('build:remove-generated');
+        invoke('deploy:vendors');
+        invoke('magento:compile');
+        invoke('magento:deploy:assets');
+        invoke('artifact:package');
+    } else {
+        throw new GracefulShutdownException("Artifact can only be built locally, you provided a non local host");
```

Comment:
I've asked Anton for clarification here about local tasks: https://github.com/deployphp/deployer/issues/2679#issuecomment-1300509342

but if there's no cleaner way I would turn this into an array of tasks without `invoke()` too (see below)